lib/chef/knife/winrm_knife_base.rb

```diff
@@ -0,0 +1,317 @@
+            queue = Queue.new
+            @winrm_sessions.each { |s| queue << s }
+            num_sessions = config[:concurrency]
+            num_targets = @winrm_sessions.length
+            num_sessions = (num_sessions.nil? || num_sessions == 0) ? num_targets : [num_sessions, num_targets].min
+
+            # These nils will kill the Threads once no more sessions are left
+            num_sessions.times { queue << nil }
+            threads = []
+            num_sessions.times do
+              threads << Thread.new do
+                while session = queue.pop
+                  run_command_in_thread(session, command)
+                end
+              end
+            end
+            threads.map(&:join)
+            @session_results
+          end
+
+          private
+
+          def run_command_in_thread(s, command)
+            @session_results << s.relay_command(command)
+          rescue WinRM::WinRMHTTPTransportError, WinRM::WinRMAuthorizationError => e
+            if authorization_error?(e)
+              unless config[:suppress_auth_failure]
+                # Display errors if the caller hasn't opted to retry
+                ui.error "Failed to authenticate to #{s.host} as #{config[:winrm_user]}"
+                ui.info "Response: #{e.message}"
+                ui.info get_failed_authentication_hint
+                raise e
+              end
+            else
+              raise e
+            end
+          end
+
+          def get_failed_authentication_hint
+            if @session_opts[:basic_auth_only]
+              FAILED_BASIC_HINT
+            else
+              FAILED_NOT_BASIC_HINT
+            end
+          end
+
+          def authorization_error?(exception)
+            exception.is_a?(WinRM::WinRMAuthorizationError) ||
+              exception.message =~ /401/
+          end
+
+          def check_for_errors!
+            @exit_code ||= 0
+            @winrm_sessions.each do |session|
+              session_exit_code = session.exit_code
+              unless success_return_codes.include? session_exit_code.to_i
+                @exit_code = [@exit_code, session_exit_code.to_i].max
+                ui.error "Failed to execute command on #{session.host} return code #{session_exit_code}"
+              end
+            end
+          end
+
+          def success_return_codes
+            # Redundant if the CLI options parsing occurs
+            return [0] unless config[:returns]
+
+            @success_return_codes ||= config[:returns].split(",").collect(&:to_i)
+          end
+
+          def session_from_list
+            @list.each do |item|
+              Chef::Log.debug("Adding #{item}")
+              @session_opts[:host] = item
+              create_winrm_session(@session_opts)
+            end
+          end
+
+          def create_winrm_session(options = {})
+            session = Chef::Knife::WinrmSession.new(options)
+            @winrm_sessions ||= []
+            @winrm_sessions.push(session)
```

Comment:
I'd probably move this to an accessor method rather than relying on the lazy initialization of the instance variable in this method.